feat: emit COURSE_CATALOG_INFO_CHANGED signal on publish

rgraber opened the pull request:
This implements <https://github.com/openedx/edx-platform/issues/30682>.

## Depends on:
- [x]  https://github.com/openedx/openedx-events/issues/82 

## Description
Send new `COURSE_CATALOG_INFO_CHANGED` signal to the event bus whenever an appropriate course_published signal is sent.

## TODO

- [x] Figure out how to create CourseCatalogData from course_id
- [x] Abstract event bus library into setting (I couldn't figure out how to do this)
    - Deferred: openedx-events will handle configurable loading later in https://github.com/openedx/openedx-events/issues/87
- [x] Append environment type to topic name
    - I think we're deferring this.
- [x] Actually send to event bus
- [x] Copy all code over to draft_and_published (and hopefully figure out how to test it)
- [x] Unit tests
- [x] Undo accidental formatting changes
- [x] Make sure course keys are actually for courses and are not LibraryLocators

## Testing instructions

* In devstack, start Kafka and the control webapp: ``make dev.up.kafka-control-center`` and watch ``make dev.logs.kafka-control-center`` until server is up and happy (may take a few minutes; wait for http://localhost:9021/clusters to come up and show a healthy cluster)
* Add `SEND_CATALOG_INFO_SIGNAL = True` to `cms/envs/private.py` (enables feature, but also causes a server restart)
* Bring up studio with make `make dev.up.studio`
* In `make studio-shell`, run `make requirements` and then `pip install 'confluent_kafka[avro,schema-registry]'`
* Run `make dev.restart-devserver.studio` to make sure new requirements are picked up
* In studio, choose a course and go to Settings > Schedule & Details
* Change the pacing, start, or end date, and save
* In the Kafka control center (localhost:9021) navigate to Topics > course-catalog-info-changed > Messages
* Use Jump to Offset -> Offset 0 to see all messages
* You should see a message with the new course details
* If you keep this page open in the control center, you should be able to see all new messages as they come in. This should happen every time you change a field or create a new course run

Commit:
feat: Emit COURSE_CATALOG_INFO_CHANGED signal on course publish

Implements https://github.com/openedx/edx-platform/issues/30682

Produce signal only once transaction for a course publish is
committed, and only for actual courses (not libraries).

- Use newer openedx-events version that has a fix for None datetime
  and that has CourseCatalogData without org, number.
- Add edx-event-bus-kafka -- specify recent version that drops
  confluent-kafka from explicit deps
- New functionality behind toggle

As per https://github.com/openedx/openedx-events/issues/88 we're going to
try explicit dependencies on implementations for now, rather than solve
all the problems we'd encounter by using private dependencies.

rgraber commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -10,6 +10,9 @@
 from django.dispatch import receiver
 from pytz import UTC
 
+from edx_event_bus_kafka.publishing.event_producer import send_to_event_bus
```
This is supposed to be configurable (ie we should be able to choose from which event bus library we're calling send_to_event_bus rather than assuming edx_event_bus_kafka).

rgraber replied:
See https://github.com/edx/internal-dockerfiles/blob/master/edx-platform-private/Dockerfile for we added a dependency *only* for edx.org (the edx-arch-experiments line) for the containerized version of edx-platform. This may or may not be applicable for the current live version of edx-platform.

timmc-edx replied:
For the moment we're likely to keep the explicit reference (under a try-import-except), pending https://github.com/openedx/openedx-events/pull/85

timmc-edx replied:
See for latest: https://github.com/openedx/openedx-events/issues/88

rgraber commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -79,6 +82,12 @@ def listen_for_course_publish(sender, course_key, **kwargs):  # pylint: disable=
+@receiver(COURSE_CATALOG_INFO_CHANGED)
+def listen_for_course_catalog_info_changed(sender, signal, **kwargs):
+    print(f"kwargs: {kwargs}")
+    send_to_event_bus(COURSE_CATALOG_INFO_CHANGED, f"course-catalog-info-changed", 'catalog_info.course_key',
```
We should really have a different topic for dev, stage, and prod environments.

timmc-edx replied:
I'm thinking that maybe the openedx-events interface can add an environment prefix/suffix.

timmc-edx replied:
(From discussion elsewhere: It may actually happen in the implementation layer instead.)

rgraber commented on cms/envs/devstack.py:
```diff
@@ -291,6 +291,11 @@ def should_show_debug_toolbar(request):  # lint-amnesty, pylint: disable=missing
 # Devstack is directly exposed to the caller
 CLOSEST_CLIENT_IP_FROM_HEADERS = []
 
+#################### Event bus backend ########################
```
These need to actually be configurable and not weirdly hardcoded

rgraber commented on requirements/common_constraints.txt:
```diff
@@ -3,6 +3,21 @@
 # See BOM-2721 for more details.
 # Below is the copied and edited version of common_constraints
 
+# This is a temporary solution to override the real common_constraints.txt
```
Oops.

rgraber commented on requirements/edx/base.in:
```diff
@@ -84,6 +84,7 @@ edx-django-sites-extensions
 edx-django-utils>=5.0.0            # Utilities for cache, monitoring, and plugins
 edx-drf-extensions
 edx-enterprise
+edx-event-bus-kafka
```
This should be moved to the private Dockerfile.

rgraber commented on xmodule/modulestore/__init__.py:
```diff
@@ -73,7 +78,8 @@ class RevisionOption:
         """
         Branch constants to use for stores, such as Mongo, that have only 2 branches: DRAFT and PUBLISHED
-        Note: These values are taken from server configuration settings, so should not be changed without alerting DevOps  # lint-amnesty, pylint: disable=line-too-long
+        Note: These values are taken from server configuration settings, so should not be changed without alerting
```
Accidental reformat, sorry!

rgraber commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +324,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
         if signal_handler and bulk_ops_record.has_publish_item:
             signal_handler.send("pre_publish", course_key=course_id)
 
+    def create_catalog_data_for_signal(self, course_id):
```
This is actual important code

rgraber commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +324,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
+        Returns:
+            CourseCatalogData
+        """
+        # Most of this is copied from CourseOverview. We cannot import it
```
Note this. Most of this code was copied because
A. we're not supposed to use CourseOverview and
B. circular import

rgraber replied:
Mostly stolen from https://github.com/openedx/edx-platform/blob/db32ff2cdf678fa8edd12c9da76a76eef0478614/openedx/core/djangoapps/content/course_overviews/models.py#L148

rgraber commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +324,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
         if self.signal_handler and bulk_ops_record.has_publish_item:
             # We remove the branch, because publishing always means copying from draft to published
             self.signal_handler.send("course_published", course_key=course_id.for_branch(None))
+            catalog_data = self.create_catalog_data_for_signal(course_id)
```
This event is only for one course but as part of bulk operations, which seems to be what happens when running basic Studio locally.

rgraber commented on xmodule/modulestore/draft_and_published.py:
```diff
@@ -136,8 +141,11 @@ def _flag_publish_event(self, course_key):
+                print(f"reload damn you")
                 # We remove the branch, because publishing always means copying from draft to published
                 self.signal_handler.send("course_published", course_key=course_key.for_branch(None))
+                catalog_info = self.create_catalog_data_for_signal(course_key)
```
This is other bit of actually important code. It can use create_catalog_data_for_signal because it inherits from the BulkOperationsMixin, which is where that method is defined. This seems to be how we handle non-bulk operations, but I couldn't figure out how to hit this code using my local Studio.

rgraber commented on xmodule/modulestore/draft_and_published.py:
```diff
@@ -7,11 +7,16 @@
-from .exceptions import ItemNotFoundError
 
 # Things w/ these categories should never be marked as version=DRAFT
+from .. import course_metadata_utils
```
Accidental comment splice

rgraber commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +324,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
         """
         if self.signal_handler and bulk_ops_record.has_publish_item:
             # We remove the branch, because publishing always means copying from draft to published
             self.signal_handler.send("course_published", course_key=course_id.for_branch(None))
```
Add a filter here to make sure course_id is not an instance of LibraryLocator before sending to the event bus

rgraber commented on xmodule/modulestore/draft_and_published.py:
```diff
@@ -136,8 +141,11 @@ def _flag_publish_event(self, course_key):
             else:
+                print(f"reload damn you")
                 # We remove the branch, because publishing always means copying from draft to published
                 self.signal_handler.send("course_published", course_key=course_key.for_branch(None))
```
Add a filter here to make sure course_id is not an instance of LibraryLocator before sending to the event bus

timmc-edx replied:
I used `course_key.is_course` based on https://github.com/openedx/opaque-keys/blob/03254a909eb428b86eda0b897e03ed6ea87aa600/opaque_keys/edx/keys.py#L25 -- seem right?

timmc-edx commented:
(Removed accidental autoformat checkins and squash-rebased to clean up branch history.)

timmc-edx commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -79,6 +82,12 @@ def listen_for_course_publish(sender, course_key, **kwargs):  # pylint: disable=
+def listen_for_course_catalog_info_changed(sender, signal, **kwargs):
+    print(f"kwargs: {kwargs}")
+    send_to_event_bus(COURSE_CATALOG_INFO_CHANGED, f"course-catalog-info-changed", 'catalog_info.course_key',
+                      {'catalog_info': kwargs['catalog_info']}, sync=True)
```
From discussion: We need to figure out why sync=true was required during testing, so we can remove it.

timmc-edx replied:
OK! Turns out the producer is getting deallocated at the end of the request. It [has a destructor](https://github.com/confluentinc/confluent-kafka-python/blob/d961259dfe25a65c23e2921be381da24eb24d253/src/confluent_kafka/src/Producer.c#L876) which [cleans up any threads](https://github.com/edenhill/librdkafka/blob/b47da0e4eec900d9cb091d2f524e957505dbb513/src/rdkafka.c#L1096) and other resources.

I experimented with event-bus-kafka and found that decorating `get_producer_for_signal` with `lru_cache` allows the event to send on this branch.

The good news is that creating an unbounded number of producers does not leak threads. The bad news is that even while iterating on this, we really can't get away with doing that, since it means producers die too quickly, and the events go with them. (I don't think `sync=True`, aka `flush()`, is a good idea in production.) I really think we need to switch to a single long-lived producer (or a small set of them).

timmc-edx replied:
Updated https://github.com/openedx/event-bus-kafka/issues/16 with new information.

timmc-edx commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +319,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
+        except ItemNotFoundError:
+            effort = None
+        return CourseCatalogData(
+            course_key=course_id.for_branch(None),
```
Does this need a comment as well, like the one in `send_bulk_published_signal`? And is it always true that this method will be called during a publish? (Maybe there needs to be a guard, or a different method name.)

timmc-edx commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +319,54 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
+        """
+        # Most of this is copied from CourseOverview. We cannot import it
+        # because it would lead to circular imports
+        course_from_store = self.get_course(course_id)
```
Where is `self.get_course` defined?

timmc-edx replied:
Turns out this is _probably_ defined on all descendents; we should consider adding an abstract method here. One wrinkle: Some subclasses call it course_id, some call it course_key. Maybe make it a positional-only param, if we can?

timmc-edx replied:
Proposal from discussion: Rather than defining `get_course` and `get_item` as abstract methods on this mixin...

1. Move the code outside of this modulestore class and into a handler for the existing `course_published` signal.
2. Update the code to call `modulestore()` to _get_ a modulestore instance and call its `get_course` method.
3. Call `CourseDetails.fetch_about_attribute` in order to get the effort information. (Might not create a circular dependency this way...)

@ormsbee Is there anything we need to worry about here with regard to transactionality? I don't know how Mongo interacts with our implicit transactions, and whether we should be 1) making sure we only build and send this event after the transaction has successfully completed (so as not to send "ghost" updates in the case of tx rollbacks), and/or 2) whether there are risks of getting the course and effort item separately (consistency issues).

timmc-edx replied:
(I've pushed a new commit, 95d3fdd, which implements this proposal.)

ormsbee replied:
There are four ModuleStore classes that are instantiated, and you can rely on `get_course` to exist in them. Please don't add an abstract method. ModuleStore has a pile of inheritance soup, but we're slowly making progress killing off the worst pieces of it via BD-13 and some new Old-Mongo-DEPR work slated to start soon.

The approach you took to move it to a handler in contentstore is fine.

In terms of transactionality–as long as you're doing it from the handler after catching a publish signal, you're probably fine. There are situations in which it can get messed up, but that affects just about everything, and I don't think you should try to solve for it in the scope of these efforts.

The one thing you will encounter a lot though is that you're going to get these signals for both draft and publish branch changes. Since you only care about things that have been published to the LMS, you might want to do something like:

https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/outlines.py#L336-L338

Also, there is a weird thing that happens to UsageKeys for changed items in a publish when you ask for the published branch in Studio. It probably doesn't affect you (since you're not really reading any UsageKeys directly), but just in case, please read: https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/outlines.py#L25-L43

timmc-edx replied:
- I've moved the signal-send to be after transaction commit, although I wonder if that deferral should actually be done by the signal-event translation layer (that we haven't written yet).
- I've used the `branch_setting` context manger you pointed out. I don't really understand all the details of branches, but I think this along with the ` course_key=course_key.for_branch(None)` should give the "published only" view of the course.
- I definitely don't know much about usage keys, but it looks like it won't affect this.

timmc-edx commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +316,52 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
+        return CourseCatalogData(
+            course_key=course_id.for_branch(None),
+            name=course_from_store.display_name,
+            org=course_from_store.location.org,
```
Raised a question about whether we want these fields: https://github.com/openedx/openedx-events/issues/72#issuecomment-1212177933

timmc-edx replied:
Fields removed in https://github.com/openedx/openedx-events/pull/95

timmc-edx commented on xmodule/modulestore/__init__.py:
```diff
@@ -313,13 +316,52 @@ def send_pre_publish_signal(self, bulk_ops_record, course_id):
+        # see CourseDetail.fetch_about_attribute
+        effort_usage_key = course_id.make_usage_key('about', 'effort')
+        try:
+            effort = self.get_item(effort_usage_key).data
```
This one is also a missing abstract method (like `get_course`).

timmc-edx commented on requirements/edx/base.in:
```diff
@@ -84,6 +84,7 @@ edx-django-sites-extensions
 edx-django-utils>=5.0.0            # Utilities for cache, monitoring, and plugins
 edx-drf-extensions
 edx-enterprise
+edx-event-bus-kafka>=0.3.0          # Kafka implementation of event bus
```
Need to double-check if this is installable on M1 Macs...

timmc-edx replied:
Dealt with via https://github.com/openedx/event-bus-kafka/pull/26

timmc-edx commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -43,6 +51,48 @@ def wrapper(*args, **kwargs):
+                enrollment_end=course.enrollment_end,
+            ),
+            effort=CourseDetails.fetch_about_attribute(course_key, 'effort'),
+            hidden=course.catalog_visibility in ['about', 'none'] or course.id.deprecated,
```
Is this `course.id` the same as the `course_key` we already have? I couldn't figure out what `get_course` even returns...

dianakhuang replied:
Yes, `course.id` should return a `CourseKey` and they should match.

timmc-edx commented on cms/djangoapps/contentstore/signals/tests/test_handlers.py:
```diff
@@ -0,0 +1,46 @@
+        )
+
+    def test_regular_course(self):
+        assert sh.create_catalog_data_for_signal(self.course_key) == self.expected_data
```
Things I might still want to test, if feasible:

- :question: Draft saves don't cause signals to be emitted (or if they do, the signal gets the old, published data)
- :heavy_check_mark: Library publishing doesn't invoke this code
- :question: If called with a course key with a branch or version (which shouldn't happen) then the emitted data still has branch=None
- :heavy_check_mark: The `emit` call causes a signal send, and doesn't blow up on None

dianakhuang commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -43,6 +51,48 @@ def wrapper(*args, **kwargs):
+        course = store.get_course(course_key)
+
+        return CourseCatalogData(
+            course_key=course_key.for_branch(None),
```
I don't think `for_branch` is necessary here.

timmc-edx replied:
Is that because `course_published` already does it, or because some other reason?

dianakhuang replied:
I think this is actually related to your comment below about how you're worried that we're getting course keys with branch versions passed around. That's not really common, but we're in modulestore land, so who knows.

dianakhuang commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -43,6 +51,48 @@ def wrapper(*args, **kwargs):
+    """
+    Given the key of a recently published course, send course data to catalog-info-changed signal.
+    """
+    if catalog_info := create_catalog_data_for_signal(course_key):
```
nit: we don't use this syntactic sugar much, and I feel like it might make the code a little harder to read, but overall, this usage is simple enough to figure out what it is.

timmc-edx replied:
I like terse syntax, but I don't have strong feelings on this one, so I'd be happy to expand it.

dianakhuang reviewed (commented):
Some initial comments.

timmc-edx commented:
(Squashed onto master; re-tested manually.)

dianakhuang commented on cms/djangoapps/contentstore/signals/handlers.py:
```diff
@@ -79,6 +146,27 @@ def listen_for_course_publish(sender, course_key, **kwargs):  # pylint: disable=
+        # FIXME: This needs to be replaced with a configurable loading
+        # mechanism built into openedx-events.
+        # See https://github.com/openedx/openedx-events/pull/85)
+        from edx_event_bus_kafka.publishing.event_producer import send_to_event_bus
```
Hmm, now that event bus kafka is a dependency, do we still want to use this pattern?

timmc-edx replied:
Mmm, we can switch that to a top-import (though we'll just need to remove it again later once it's configurable).

timmc-edx replied:
Should be fixed now.

dianakhuang commented on cms/djangoapps/contentstore/signals/tests/test_handlers.py:
```diff
@@ -0,0 +1,66 @@
+    def test_disabled(self, mock_signal):
+        """When toggle is disabled, don't send."""
+        sh.emit_catalog_info_changed_signal(self.course_key)
+        mock_signal.send_event.assert_not_called()
```
I would really like to see tests of course_published that verify the catalog info changed signal is also fired if at all possible.

timmc-edx replied:
I tried making a test for that, but ran into some trouble with getting the transaction-on-commit to play nice with unit tests. I can take another crack at it.

dianakhuang replied:
Bah, that sucks, but if you can't get it to work, I guess we can kick that can down the road.

timmc-edx replied:
I could probably mock out the on-commit and confirm that it's _called_, though... I'll give it a try.

timmc-edx replied:
I pushed something up. It's a little weird, but I guess it works?

It includes some new code for SwitchedSignal that I was surprised didn't already exist, but maybe I'm missing something: https://github.com/openedx/edx-platform/pull/30805/commits/ee5681857ea5c8aad0280de501434d4ab282258a#diff-29c3ba3788011c17dfcd15a34f80efbeb2cf19a09c204e7ede94dc42fde671ccR92

timmc-edx committed:
fixup! Move variable out of try block (thanks, linter!)

timmc-edx committed:
fixup! Turn edx-event-bus-kafka constraint into an exact-version pin

...and upgrade it to a version that includes the auth setting fix and the
multi-producer caching tweak.

timmc-edx committed:
Merge branch 'master' into rsgraber/send-new-signal
